admin/release.py

```diff
@@ -67,3 +92,212 @@ def make_rpm_version(flocker_version):
+        if get_doc_version(flocker_version) != doc_version:
+            raise NotTagged
+    configuration = DOCUMENTATION_CONFIGURATIONS[environment]
+
+    dev_prefix = '%s/' % (flocker_version,)
+    version_prefix = 'en/%s/' % (doc_version,)
+
+    is_dev = not is_release(doc_version)
+    if is_dev:
+        stable_prefix = "en/devel/"
+    else:
+        stable_prefix = "en/latest/"
+
+    # Get the list of keys in the new documentation.
+    new_version_keys = yield Effect(
+        ListS3Keys(bucket=configuration.dev_bucket,
+                   prefix=dev_prefix))
+    # Get the list of keys already existing for the given version.
+    # This should only be non-empty for documentation releases.
+    existing_version_keys = yield Effect(
+        ListS3Keys(bucket=configuration.documentation_bucket,
+                   prefix=version_prefix))
+
+    # Copy the new documentation to the documentation bucket.
+    yield Effect(
+        CopyS3Keys(source_bucket=configuration.dev_bucket,
+                   source_prefix=dev_prefix,
+                   destination_bucket=configuration.documentation_bucket,
+                   destination_prefix=version_prefix,
+                   keys=new_version_keys))
+
+    # Delete any keys that aren't in the new documentation.
+    yield Effect(
+        DeleteS3Keys(bucket=configuration.documentation_bucket,
+                     prefix=version_prefix,
+                     keys=existing_version_keys - new_version_keys))
+
+    # Update the redirect for the stable URL (en/latest/ or en/devel/)
+    # to point to the new version. Returns the old target.
+    old_prefix = yield Effect(
+        UpdateS3RoutingRule(bucket=configuration.documentation_bucket,
+                            prefix=stable_prefix,
+                            target_prefix=version_prefix))
+
+    # If we have changed versions, get all the keys from the old version
+    if old_prefix:
+        previous_version_keys = yield Effect(
+            ListS3Keys(bucket=configuration.documentation_bucket,
+                       prefix=old_prefix))
+    else:
+        previous_version_keys = set()
+
+    # The changed keys are the new keys, the keys that were deleted from this
+    # version, and the keys for the previous version.
+    changed_keys = (new_version_keys |
+                    existing_version_keys |
+                    previous_version_keys)
+
+    # S3 serves /index.html when given /, so any changed /index.html means
+    # that / changed as well.
+    # Note that we check for '/index.html' but remove 'index.html'
+    changed_keys |= {key_name[:-len('index.html')]
+                     for key_name in changed_keys
+                     if key_name.endswith('/index.html')}
+
+    # Always update the root.
+    changed_keys |= {''}
+
+    # The full paths are all the changed keys under the stable prefix, and
+    # the new version prefix. This set is slightly bigger than necessary.
```

Comment:
The reason I didn't explain more than I did is that it isn't straightforward, and doesn't really matter. For example, the keys that are only in the version that was previous `/latest/` only needs to be invalidated for `/latest/` and not `/<version>/`. There may be others, but it doesn't really hurt to invalidate extra ones.